serializers/src/main/java/io/pravega/schemaregistry/serializers/SerializerConfig.java

```diff
@@ -0,0 +1,278 @@
+     */
+    @Getter(AccessLevel.NONE)
+    private final SchemaRegistryClient registryClient;
+    /**
+     * Flag to tell the serializer if the schema should be automatically registered before using it in {@link io.pravega.client.stream.EventStreamWriter}.
+     * It is recommended to register keep this flag as false in production systems and manage schema evolution explicitly and
+     * in lockstep with upgrade of existing pravega client applications.
+     */
+    private final boolean registerSchema;
+    /**
+     * Flag to tell the serializer if the codec should be automatically registered before using the serializer in
+     * {@link io.pravega.client.stream.EventStreamWriter}.
+     * It is recommended to register keep this flag as false in production systems and manage codecTypes used by writers explicitly
+     * so that readers are aware of encodings used.
+     */
+    private final boolean registerCodec;
+    /**
+     * Codec to use for encoding events after serializing them.
+     */
+    private final Encoder encoder;
+    /**
+     * Function that should be applied on serialized data read from stream. This is invoked after reading the codecType
+     * from {@link EncodingInfo} and using the codec type read from it.
+     * It should return the decoded data back to the deserializer.
+     * Use {@link SerializerConfigBuilder#decoder(String, Decoder)} to add decoders. 
+     * Any number of decoders can be added. 
+     */
+    private final Decoders decoders;
+    /**
+     * Tells the deserializer that if supplied decoder codecTypes do not match group codecTypes then fail and exit upfront.
+     */
+    private final boolean failOnCodecMismatch;
+    /**
+     * Flag to tell the serializer/deserializer if the group should be created automatically.
+     * It is recommended to register keep this flag as false in production systems and create groups and add schemas
+     */
+    @Getter(AccessLevel.NONE)
+    private final GroupProperties createGroup;
+    /**
+     * Flag to tell the serializer/deserializer if the encoding id should be added as a header with each event.
+     * By default this is set to true. If users choose to not add the header, they should do so in all their writer and 
+     * reader applications for the given stream. 
+     *
+     * Adding the event header is a requirement for following cases: 
+     * If {@link SerializationFormat#Avro} is chosen for a group, the event header cannot be false.
+     * If streams can have multiple types of events, this cannot be false.
+     * If streams can multiple formats of events, this cannot be false.
+     */
+    private final boolean writeEncodingHeader;
+
+    private SerializerConfig(String groupId, String namespace, SchemaRegistryClientConfig config, SchemaRegistryClient client,
+                             boolean registerSchema, boolean registerCodec, Codec encoder, Decoders decoders, boolean failOnCodecMismatch,
```

Comment:
I'm confused with the terminology used here. The term `Codec` is supposed to refer to both encoder and decoder, but here codec is only the encoder and the decoder is supposed to be passed separately as a list of `Decoders`. It is a bit confusing who this encoding and decoding needs to happen.

Reply:
For encoding the data, each writer uses at most one `encoding function`. 

For decoding, there needs to be multiple decoders, typically a decoder function for each of registered codecTypes so that the deserializer knows how to decode the data.